driver/ixsystems/common.py

```diff
@@ -409,41 +409,60 @@ def _create_volume_from_snapshot(self, name, snapshot_name,
         except Exception as e:
             raise FreeNASApiError('Unexpected error', e)
 
+    def _system_version(self):
+            LOG.debug('_update_volume_stats start /system/version request')
+            request_urn = ("/system/version")
+            self.handle.set_api_version('v2.0')
+            versionresult=""
+            try:
+                versionret = self.handle.invoke_command(FreeNASServer.SELECT_COMMAND,
+                                                request_urn, None)
+                LOG.debug('_update_volume_stats start /system/version response: %s',versionret)
+                versionresult = json.loads(versionret['response'])
+                LOG.debug('_update_volume_stats /system/version response : %s', versionresult)
+            except Exception as e:
+                raise FreeNASApiError('Unexpected error', e)
+            finally:
+                return str(versionresult)
+            
     def _update_volume_stats(self):
-        """Retrieve stats info from volume group.
-
-           REST API: $ GET /pools/mypool "size":95,"allocated":85,
-        """
-        # HACK: for now, use an API v1.0 call to get
-        # these stats until available in v2.0 API
-        self.handle.set_api_version('v1.0')
-        request_urn = ('%s/%s/') % (
-            '/storage/volume',
-            self.configuration.ixsystems_datastore_pool)
-        LOG.debug('_update_volume_stats request_urn : %s', request_urn)
-        ret = self.handle.invoke_command(FreeNASServer.SELECT_COMMAND,
-                                         request_urn, None)
-        LOG.debug("_update_volume_stats response : %s", json.dumps(ret))
         data = {}
-        data["volume_backend_name"] = self.backend_name
-        data["vendor_name"] = self.vendor_name
-        data["driver_version"] = self.VERSION
-        data["storage_protocol"] = self.storage_protocol
-        data['total_capacity_gb'] = ix_utils.get_size_in_gb(
-            json.loads(ret['response'])['avail'] +
-            json.loads(ret['response'])['used'])
-        data['free_capacity_gb'] = ix_utils.get_size_in_gb(
-            json.loads(ret['response'])['avail'])
-        data['reserved_percentage'] = (
-            self.configuration.ixsystems_reserved_percentage)
-        data['reserved_percentage'] = 0
-        data['QoS_support'] = False
+        # Implementation for TrueNAS 12.0 upwards on API V2.0
+        # If user are connecting to FreeNAS report error
+        if self._system_version().find("FreeNAS")>=0:
```

Comment:
How can this ever return FreeNAS if _system_version is set to v2.0 and the fallback is an empty string?

Reply:
Hi William, I've updated the code contains extra error handling for version fault back value, error handling and some extra comments in code for system that does not even has v2.0 API. Please review. Thank you!